### PR TITLE
LPS-197665 | Allow users to edit schema properties

### PR DESCRIPTION
**Pull request by Hijky**
**Background:**
Add a testcase to allow users to edit schema properties

**Test Plan**
PropertiesSectionIsPresentForSchema
When in Schemas of APIApplication a 'University' Schema based on University object created
Then Properties section appears in Schemas > University

PropertiesSidebarIsPresentForSchema
When going to Properties section in Schemas > University
Then Properties sidebar present

ObjectFieldsArePresentInPropertiesSidebar
When going to Properties section in Schemas > University
Then University fields are present in Properties sidebar

RelatedObjectsOptionIsPresentInPropertiesSidebar
When going to Properties section in Schemas > University
Then 'View Related Objects' option is present in Properties sidebar

FieldsOfRelatedObjectsAreAccessibleFromPropertiesSidebar
When clicking 'View Related Objects' option in Properties sidebar
Then all fields of Address object definition are present

DroppedIntoPropertiesListRelatedObjectFieldIsDisabledInPropertiesSidebar
And Given fields of Address object definition present through View Related Objects'
When dragging 'External Reference Code' of Address to 'Drop properties...' box
Then 'External Reference Code' is placed in Properties

CanDragAndDropPropertiesFromSidebarToList
When dragging the 'name' field from University section of Properties sidebar to 'Drop properties...' box
Then 'name' field is placed in Properties

CanSearchThroughPropertiesFromTheList
And Given 'name', 'External Reference Code', 'Creation Date', 'ID' fields placed in Properties
When searching property 'External Reference Code' with Search box
Then only 'External Reference Code' present in Properties list

**Jira ticket:**
https://liferay.atlassian.net/browse/LPS-197665

### CODE REVIEW

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/APIBuilderUI.macro [hunk 1]
```diff
@@ -395,6 +391,17 @@ definition {
 			value = ${tabEntity});
 	}
 
+	@summary = "Default summary"
+	macro gotoRelatedObjectsProperties(relatedObjectName = null) {
```
Comment (magjed4289) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/APIBuilderUI.macro:395]: 👍🏻

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+					value1 = ${fieldName});
+			}
+		}
+	}
+
+	@priority = 5
+	test ObjectFieldsArePresentInPropertiesSidebar {
+		property portal.acceptance = "true";
+
+		task ("Then University fields are present in Properties sidebar") {
```
Comment (magjed4289) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:175]: Where is a `When` step for this test?

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+					value1 = ${fieldName});
+			}
+		}
+	}
+
+	@priority = 5
+	test PropertiesSectionIsPresentForSchema {
+		property portal.acceptance = "true";
+
+		task ("Then Properties section appears in Schemas > University") {
```
Comment (magjed4289) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:191]: same here

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+				locator1 = "Button#BUTTON_WITH_VALUE",
+				value = "Properties");
+		}
+	}
+
+	@priority = 5
+	test PropertiesSidebarIsPresentForSchema {
+		property portal.acceptance = "true";
+
+		task ("Then Properties sidebar present") {
```
Comment (magjed4289) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:202]: same here

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+				key_header = "Properties",
+				locator1 = "ProcessBuilderKaleoDesignerReact#SIDEBAR_HEADER");
+		}
+	}
+
+	@priority = 4
+	test RelatedObjectsOptionIsPresentInPropertiesSidebar {
+		property portal.acceptance = "true";
+
+		task ("Then 'View Related Objects' option is present in Properties sidebar") {
```
Comment (magjed4289) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:213]: same here

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+		}
+
+		task ("And Given 'University' Schema based on University object created") {
+			SchemaAPI.createAPISchema(
+				externalReferenceCode = "My-app",
+				mainObjectDefinitionERC = "university",
+				name = "University");
+		}
+
+		task ("When going to Properties section in Schemas > University") {
```
Comment (magjed4289) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:47]: When step cannot go in Background, it makes test hard to read

**Review by magjed4289 — CHANGES_REQUESTED** [review]

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+		if (${testLiferayVirtualInstance} == "true") {
+			PortalInstances.tearDownCP();
+		}
+	}
+
+	@priority = 4
+	test CanDragAndDropPropertiesFromSidebarToList {
+		property portal.acceptance = "true";
+
+		task ("And Given 'University' Schema based on University object created") {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:57]: We can move the 'And Given' into Setup and leave the 'When' step in every case.

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+		}
+
+		task ("When searching property 'External Reference Code' with Search box") {
+			Type(
+				locator1 = "TextInput#SEARCH",
+				value1 = "External Reference Code");
+		}
+
+		task ("Then only 'External Reference Code' present in Properties list") {
+			AssertElementNotPresent(
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:118]: We can use for to assert not present here

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+				key_commentValue = "Create Date",
+				locator1 = "PublicationsChanges#COMMENT_VALUE",
+				value1 = "Create Date");
+
+			AssertElementNotPresent(
+				key_commentValue = "ID",
+				locator1 = "PublicationsChanges#COMMENT_VALUE",
+				value1 = "ID");
+
+			AssertTextEquals(
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:133]: We can use AssertElementPresent()  and remove the unused `value1`

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+				title = "My-app");
+
+			APIBuilderUI.gotoRelatedObjectsProperties(relatedObjectName = "Postal Address");
+		}
+
+		task ("Then all fields of Address object definition are present") {
+			var fieldNameList = "Author,Create Date,External Reference Code,ID,Modified Date,Status,Name,Street 1,UniversityAddresses";
+
+			for (var fieldName : list ${fieldNameList}) {
+				AssertTextEquals(
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:200]: Same here

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+			APIBuilderUI.goToEditRelatedEntryInEditApplication(
+				entity = "Schemas",
+				name = "University",
+				tabEntity = "Properties",
+				title = "My-app");
+
+			APIBuilderUI.gotoRelatedObjectsProperties(relatedObjectName = "Postal Address");
+		}
+
+		task ("When dragging 'External Reference Code' of Address to 'Drop properties' box") {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:161]: @magjed4289 Right now we only support click on a field to add a property, drag and drop to create a property is not supported. Should we update all "drag" into "Add" or "Create" in all task(steps) and case name? cc @Hijky

Comment (Hijky) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[UI]DefineAPISchemasWithinAnAPIGroup/AllowUsersToEditSchemaProperties.testcase:161]: @magjed4289 I have modified the 'drag' to 'add' and the case name 'CanDragAndDropPropertiesFromSidebarToList' to 'CanAddPropertiesFromSidebarToList'@cc @yanan

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-builder-functional:nocompile

**Review by magjed4289 — APPROVED** [review]

**Comment by Hijky** [comment]
The failures are flaky and not related to PR.

**Comment by yanan** [comment]
ci:forward

**Pull request closed without merging**